Change OpenLicence validation message to get acceptable values dynamically

tylerjmchugh opened the pull request:
Currently the validation message for licenses has the accepted values hardcoded in the message. These accepted values should be retrieved dynamically from the thesaurus. This will allow for the implementation of custom licences.

Current messages:
![image](https://github.com/user-attachments/assets/dbff308f-2c14-408c-9815-9aaa2e9d7926)

New messages: (`DFO Internal Restricted License` is an example of a custom license)
![image](https://github.com/user-attachments/assets/c02dbd73-498b-4b48-a810-de8acc7ac1c1)

This PR aims to addresses this issue by making the validation message more generic and by removing the hardcoded values from the message.

tylerjmchugh committed:
Change OpenLicence validation message to get acceptable values dynamically

tylerjmchugh committed:
Update messages to display main language values in non-multilingual and alt language values in multilingual

tylerjmchugh committed:
Update multilingual schematron to check both languages

tylerjmchugh committed:
Update non-multilingual schematron to display a more generic message

tylerjmchugh committed:
Update non-multilingual messages to exclude the accepted values

tylerjmchugh committed:
Fix typo

josegar74 reviewed (commented):
I've tested the pull request and with an invalid value it reports the error with the list of licenses, but the English validation message shows the French license values and the French validation messages shows the English license values. See below.

**Also with a valid license shows the following:**

```
An English supported license is required. Accepted values: Licence du gouvernement ouvert - Canada (https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada) or Licence du gouvernement ouvert - Alberta (https://open.alberta.ca/licence) or Licence du gouvernement ouvert - Colombie-Britannique (https://www2.gov.bc.ca/gov/content/data/open-data/open-government-licence-bc) or Licence du gouvernement ouvert – Terre-Neuve-et-Labrador ([https://opendata.gov.nl.ca/public/opendata/page/?page-id=licence)'](https://opendata.gov.nl.ca/public/opendata/page/?page-id=licence)%27) or Licence du gouvernement ouvert – Nouvelle-Écosse (https://novascotia.ca/opendata/licence.asp) or Licence du gouvernement ouvert – Ontario (https://www.ontario.ca/fr/page/licence-du-gouvernement-ouvert-ontario) or Licence du gouvernement ouvert – Île-du-Prince-Édouard (https://www.princeedwardisland.ca/fr/information/finances/licence-du-gouvernement-ouvert-ile-du-prince-edouard)
```

josegar74 commented on src/main/plugin/iso19139.ca.HNAP/schematron/schematron-rules-multilingual.sch:
```diff
@@ -68,6 +68,24 @@
+    <xsl:variable name="open-license-list" select="document(concat('file:///', replace(concat($thesaurusDir, '/external/thesauri/theme/GC_Open_Licenses.rdf'), '\\', '/')))"/>
+
+    <xsl:variable name="v">
+      <xsl:for-each select="$open-license-list//rdf:Description/ns2:prefLabel[@xml:lang=$altLanguage2char]">
```
`$altLanguage2char` is fixed, so this function will always return the values in the alternative language, should not be a parameter of the function to send the proper value where is called?

josegar74 replied:
The code can be similar to this other one: https://github.com/metadata101/iso19139.ca.HNAP/pull/394/files#diff-7c8ff55e3e815d4f6aa802798f973211c9eab945cc11091e8795758a6b60a1deR56

josegar74 commented on src/main/plugin/iso19139.ca.HNAP/schematron/schematron-rules-multilingual.sch:
```diff
@@ -392,22 +410,32 @@
         |//*[@gco:isoType='gmd:MD_DataIdentification']
         |//*[@gco:isoType='srv:SV_ServiceIdentification']">
 
+      <sch:let name="openLicenseList" value="geonet:openLicenseList($thesaurusDir)"/>
```
See the comment above about `$altLanguage2char` hardcoded in `geonet:openLicenseList`. I think the language should be a parameter to the function and call it twice to retrieve the list of values in English and French, adapting the code below to use either of the 2 lists depending on the languages.

ianwallen commented on src/main/plugin/iso19139.ca.HNAP/schematron/schematron-rules-multilingual.sch:
```diff
@@ -68,6 +68,24 @@
+    <xsl:variable name="open-license-list" select="document(concat('file:///', replace(concat($thesaurusDir, '/external/thesauri/theme/GC_Open_Licenses.rdf'), '\\', '/')))"/>
+
+    <xsl:variable name="v">
+      <xsl:for-each select="$open-license-list//rdf:Description/ns2:prefLabel[@xml:lang=$altLanguage2char]">
```
I would expect $altLanguage2char to be replaced by a parameter supplied to the function to indicate the language of the list of values being supplied. Otherwise it seems like you will always get the alternate languages as a list.

ianwallen commented on src/main/plugin/iso19139.ca.HNAP/schematron/schematron-rules-multilingual.sch:
```diff
@@ -392,22 +410,32 @@
 
+      <sch:let name="openLicenseList" value="geonet:openLicenseList($thesaurusDir)"/>
+      <sch:let name="locMsgMain" value="geonet:appendLocaleMessage($loc/strings/*[name() = concat('OpenLicense', $mainLanguageText)], $openLicenseList)"/>
+      <sch:let name="locMsgAlt" value="geonet:appendLocaleMessage($loc/strings/*[name() = concat('OpenLicense', $altLanguageText)], $openLicenseList)"/>
```
Shouldn't $openLicenseList be the alt language list?  Seems like you are supplying the same list for both main and alt error message?

tylerjmchugh committed:
Update openLicense message to display values in the correct language

tylerjmchugh commented:
Updated to display values in the correct language. Testing results:

- Metadata with English content as the main language and French as the alt language
	- Test with English Interface.
		- Test English Missing and that the error message is English with English list of values. [PASS]
		- Test French Missing and that the error message is English with French list of values. [PASS]
	- Test with French Interface.
		- Test English Missing and that the error message is French with English list of values. [PASS]
		- Test French Missing and that the error message is French with French list of values. [PASS]
- Metadata with English content as the alt language and French as the main language
	- Test with English Interface.
		- Test English Missing and that the error message is English with English list of values. [PASS]
		- Test French Missing and that the error message is English with French list of values. [PASS]
	- Test with French Interface.
		- Test English Missing and that the error message is French with English list of values. [PASS]
		- Test French Missing and that the error message is French with French list of values. [PASS]
		
Also fixed the issue with the English validation message showing when the record is valid.

josegar74 reviewed (approved):
Tested the changes and works fine.
